can/io/blf.py

```diff
@@ -0,0 +1,263 @@
+                                      arbitration_id=can_id & 0x1FFFFFFF,
+                                      extended_id=bool(can_id & CAN_MSG_EXT),
+                                      is_remote_frame=bool(flags & REMOTE_FLAG),
+                                      dlc=dlc,
+                                      data=can_data[:dlc])
+                        msg.channel = channel
+                        yield msg
+                    elif obj_type == CAN_ERROR:
+                        channel, length = CAN_ERROR_STRUCT.unpack(obj_data)
+                        msg = Message(timestamp=timestamp, is_error_frame=True)
+                        msg.channel = channel
+                        yield msg
+                    pos += obj_size
+                    # Add padding bytes
+                    pos += obj_size % 4
+                # Save remaing data that could not be processed
+                tail = data[pos:]
+        self.fp.close()
+
+
+class BLFWriter(Listener):
+    """
+    Logs CAN data to a Binary Logging File compatible with Vector's tools.
+    """
+
+    #: Max log container size of uncompressed data
+    MAX_CACHE_SIZE = 0x20000
+
+    #: ZLIB compression level
+    COMPRESSION_LEVEL = 7
+
+    def __init__(self, filename, channel=1):
+        self.fp = open(filename, "wb")
+        self.channel = channel
+        # Header will be written after log is done
+        self.fp.write(b"\x00" * FILE_HEADER_STRUCT.size)
+        self.cache = []
+        self.cache_size = 0
+        self.count_of_objects = 0
+        self.uncompressed_size = FILE_HEADER_STRUCT.size
+        self.start_timestamp = None
+        self.stop_timestamp = None
+
+    def on_message_received(self, msg):
+        if self.start_timestamp is None:
+            self.start_timestamp = msg.timestamp
+        self.stop_timestamp = msg.timestamp
+        timestamp = int((msg.timestamp - self.start_timestamp) * 1000000000)
+        if not msg.is_error_frame:
+            obj_size = OBJ_HEADER_STRUCT.size + CAN_MSG_STRUCT.size
+            header = OBJ_HEADER_STRUCT.pack(
+                b"LOBJ", OBJ_HEADER_STRUCT.size, 1, obj_size, CAN_MESSAGE,
+                2, 0, timestamp)
+            flags = REMOTE_FLAG if msg.is_remote_frame else 0
+            arb_id = msg.arbitration_id
+            if msg.id_type:
+                arb_id |= CAN_MSG_EXT
+            data = CAN_MSG_STRUCT.pack(self.channel, flags, msg.dlc, arb_id,
+                                       bytes(msg.data))
+        else:
+            obj_size = OBJ_HEADER_STRUCT.size + CAN_ERROR_STRUCT.size
+            header = OBJ_HEADER_STRUCT.pack(
+                b"LOBJ", OBJ_HEADER_STRUCT.size, 1, obj_size, CAN_ERROR,
+                2, 0, timestamp)
+            data = CAN_ERROR_STRUCT.pack(self.channel, 0)
+        self._add_data(header + data)
+
+    def log_event(self, message, timestamp=None):
+        """Add an arbitrary message to the log file."""
```

Comment:
Since we already deal in "messages" I'd change this parameter to `text`. As it is a public api could you add the expected types to the docstring too.